File: Upcoder/Lý thuyết đồ thị/Duyệt theo chiều rộng/BFS.cs
using System;
using System.Collections.Generic;

namespace BFS
{
    public class Program
    {
        static int nE, nV, S;
        static List<int>[] adj;
        static bool[] visited;

        public static void Main(string[] args)
        {
            Input();
            BFS(S);
        }

        static void Input()
        {
            string[] input = Console.ReadLine().Split(' ');
            nE = int.Parse(input[0]);
            nV = int.Parse(input[1]);
            S = int.Parse(input[2]);

            adj = new List<int>[nE + 1];
            for (int i = 1; i <= nE; i++)
            {
                adj[i] = new List<int>();
            }

            for (int i = 0; i < nV; i++)
            {
                string[] edge = Console.ReadLine().Split(' ');
                int x = int.Parse(edge[0]);
                int y = int.Parse(edge[1]);
                adj[x].Add(y);
                adj[y].Add(x);
            }

            visited = new bool[nE + 1];
        }

        static void BFS(int S)
        {
            Queue<int> queue = new Queue<int>();
            queue.Enqueue(S);
            visited[S] = true;

            while (queue.Count > 0)
            {
                int v = queue.Dequeue();
                Console.Write(v + " ");

                foreach (int x in adj[v])
                {
                    if (!visited[x])
                    {
                        queue.Enqueue(x);
                        visited[x] = true;
                    }
                }
            }
        }
    }
}

File: Upcoder/Lý thuyết đồ thị/Duyệt theo chiều rộng/BFS.py
from collections import deque

def input_data():
    nE, nV, S = map(int, input().split())
    adj = [[] for _ in range(nV + 1)]

    for _ in range(nE):
        x, y = map(int, input().split())
        adj[x].append(y)
        adj[y].append(x)
    
    return adj, S, nV

def BFS(S, adj, nV):
    visited = [False] * (nV + 1)
    queue = deque([S])
    visited[S] = True
    
    while queue:
        v = queue.popleft()
        print(v, end=" ")
        
        for x in adj[v]:
            if not visited[x]:
                queue.append(x)
                visited[x] = True

if __name__ == "__main__":
    adj, S, nV = input_data()
    BFS(S, adj, nV)
